### PR TITLE
kube-sqs-autoscaler

### PR DESCRIPTION
**Pull request by andrewsykim**
Added unit tests and updated the README. I'll add vendors afterwards to keep the PR clean.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+.PHONY clean build push
+
+IMAGE=wattpad/kube-sqs-autoscaler
+VERSION=v1.0
+
+clean:
+    rm -f kube-sqs-autoscaler
+
+build: clean
+
```
Comment (pbitty) [inline Makefile]: Where's the rest?  :)

File: scale/scale.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	return &PodAutoScaler{
+		Client:     k8sClient,
+		Min:        min,
+		Max:        max,
+		Deployment: kubernetesDeploymentName,
+	}
+}
+
+func (p *PodAutoScaler) ScaleUp() error {
+	// TODO: use the same namespace that kube-sqs-autoscaler is using
```
Comment (pbitty) [inline scale/scale.go]: The Downward API can help here.  We can pass in a `MY_NAMESPACE` env var like we do in the nginx-ingress-controller.

File: main.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	scaleDownMessages   = flag.Int("scale-down-messages", 10, "Number of messages required to scale down")
+	maxPods             = flag.Int("max-pods", 5, "Max pods that kube-sqs-autoscaler can scale")
+	minPods             = flag.Int("max-pods", 1, "Min pods that kube-sqs-autoscaler can scale")
+	awsRegion           = flag.String("aws-region", "us-east-1", "Your AWS region")
+
+	sqsQueueUrl              = flag.String("sqs-queue-url", "", "The sqs queue url")
+	kubernetesDeploymentName = flag.String("kubernetes-deployment", "", "Kubernetes Deployment to scale. This field is required")
+)
+
+func Run(p *scale.PodAutoScaler, sqs *sqs.SqsClient) {
```
Comment (pbitty) [inline main.go]: I wonder if we should be testing this method, given this is the core logic, the "auto" part of the autoscaler.  Since the external pieces are mocked, we're most of the way there.

**Review by pbitty — COMMENTED** [review]
Review done.  Suggest we test the `Run` logic.

File: scale/scale.go (hunk 2 above)
Comment (pbitty) [reply scale/scale.go]: This isn't hard to do, just another flag and a env var in the Pod spec.  Might as well throw it in there and save the work later, what do you think?

File: Makefile (hunk 1 above)
Comment (andrewsykim) [reply Makefile]: oops, forgot to push that, its pushed now

File: scale/scale.go (hunk 2 above)
Comment (andrewsykim) [reply scale/scale.go]: 👍

File: main.go (hunk 3 above)
Comment (andrewsykim) [reply main.go]: hmm I was thinking we can do e2e testing using minikube like we did for nginx-ingress-controller but now that I think about it you're probably right we can probably mock that too. Let me take a stab at that

**Comment by andrewsykim** [comment]
@pbitty added a test for `Run()`, it involved having to duplicate some of the test code from other the scale and sqs package since they types/funcs in `*_test.go` can't be imported.

**Comment by pbitty** [comment]
@andrewsykim  cool.   Sucks about the duplication.  Maybe @jharlap has suggestions around that.

**Comment by andrewsykim** [comment]
yeah I asked him, he said if its not too much code and you're not duplicating it more than 3 times its generally okay to duplicate for testing. Otherwise you should put test specific features into a separate package

File: README.md [hunk 4]
```diff
@@ -1,2 +1,77 @@
...
+
+## Setting up
+Setting up kube-sqs-autoscaler requires two steps:
+1) Deploying it as an incluster service in your cluster
+2) Adding AWS permissions so it can read the number of messages in your queues.
+
+### Deploying kube-sqs-autoscaler
+Deployin kube-sqs-autoscaler should be as simple as applying this deployment:
+```
+---
```
Comment (pbitty) [inline README.md]: If you follow the triple-backticks with the keyword `yaml`, the yaml will look nicer on Github.

Comment (pbitty) [inline README.md]: If you follow the triple-backticks with the keyword `yaml`, the yaml will look nicer on Github.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,77 @@
...
+            readOnly: true
+    volumes:
+     - name: ssl-certs
+       hostPath:
+         path: "/etc/ssl/certs/ca-certificates.crt"
+```
+
+### Permissions
+Next you want to attach this policy so kube-sqs-autoscaler can retreive SQS attributes:
+```
```
Comment (pbitty) [inline README.md]: Same for `json`

File: scale/scale.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	"os"
+
+	"github.com/pkg/errors"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+var podNamespace = os.Getenv("POD_NAMESPACE")
```
Comment (pbitty) [inline scale/scale.go]: This is the only runtime parameter that is configurable via an env var.  I suggest moving it to a CLI flag and doing interpolation via the Pod spec.  ie:

```
/kube-sqs-autoscaler --kubernetes-namespace=$(POD_NAMESPACE)
```

That way all options are in the same place.  Just more of an operational thing.

**Comment by pbitty** [comment]
Review done.  Looks great!  Minor comment re: ENV_VAR, then :shipit: from me.

**Comment by pbitty** [comment]
Nice work @andrewsykim

File: scale/scale.go (hunk 6 above)
Comment (andrewsykim) [reply scale/scale.go]: makes sense

Comment (andrewsykim) [reply scale/scale.go]: done

File: README.md (hunk 5 above)
Comment (andrewsykim) [reply README.md]: done

File: README.md (hunk 4 above)
Comment (andrewsykim) [reply README.md]: done

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,3 @@
+FROM alpine:3.4
```
Comment (jharlap) [inline Dockerfile:1]: just to bug paulo: have you considered `FROM scratch`? ;P

File: Makefile [hunk 8]
```diff
@@ -0,0 +1,16 @@
+.PHONY clean compile build push
+
+IMAGE=wattpad/kube-sqs-autoscaler
+VERSION=v1.0
+
+clean:
+    rm -f kube-sqs-autoscaler
+
+compile: clean
```
Comment (jharlap) [inline Makefile]: might want to add a `test` target - your makefile effectively becomes a quick intro to anyone who wants to work on the codebase

File: README.md [hunk 9]
```diff
@@ -1,2 +1,78 @@
 # kube-sqs-autoscaler
-Kubernetes pod autoscaler based on queue size in AWS SQS
+Kubernetes pod autoscaler based on queue size in AWS SQS. It periodically retrieves the number of messages in your queue and scales pods accordingly.
+
+## Setting up
+Setting up kube-sqs-autoscaler requires two steps:
+1) Deploying it as an incluster service in your cluster
+2) Adding AWS permissions so it can read the number of messages in your queues.
```
Comment (jharlap) [inline README.md:7]: this feels like smell. the same way my service owns its datastore and other things shouldn't peek inside, it effectively owns its queue and other things shouldn't peek inside. 👍 for the quick solution, but we should probably think about a better way to do this... for example, maybe queue consumer services should always have an endpoint that lets you query for an estimate of backlog size

File: README.md [hunk 10]
```diff
@@ -1,2 +1,78 @@
...
+
+### Deploying kube-sqs-autoscaler
+Deployin kube-sqs-autoscaler should be as simple as applying this deployment:
```
Comment (jharlap) [inline README.md:10]: Deployin' is so hip! 😜

File: sqs/sqs.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sqs"
+
+	"github.com/pkg/errors"
+)
+
+// for testing purposes
```
Comment (jharlap) [inline sqs/sqs.go]: don't think you need to specify this - it's just good go practice for how you do dependency injection in a weak-coupling way.

File: sqs/sqs.go [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+type SQS interface {
+	GetQueueAttributes(*sqs.GetQueueAttributesInput) (*sqs.GetQueueAttributesOutput, error)
+	// only implemented on unit tests
+	SetQueueAttributes(*sqs.SetQueueAttributesInput) (*sqs.SetQueueAttributesOutput, error)
```
Comment (jharlap) [inline sqs/sqs.go]: why is it part of the interface you consume if you never call it?

File: README.md [hunk 13]
```diff
@@ -1,2 +1,78 @@
...
+  template:
+    metadata:
+      labels:
+        app: kube-sqs-autoscaler
+    spec:
+      containers:
+      - name: kube-sqs-autoscaler
+        image: wattpad/kube-sqs-autoscaler:v1.0
+        command:
+          - /kube-sqs-autoscaler
```
Comment (jharlap) [inline README.md:33]: i wonder if it's more docker™ to make kube-sqs-autoscaler be the entry point and then you can just pass all the args you want without having to override the command itself. or to go 12FA and not take command line arguments, only env vars

Comment (jharlap) [inline README.md]: or to go 12FA and not take command line arguments, only env vars

File: README.md (hunk 9 above)
Comment (pbitty) [reply README.md:7]: Interesting.  I like it.  That would also simplify IAM permissions.

Comment (andrewsykim) [reply README.md:7]: Agreed, in the future when we have custom metrics setup it should roughly look like that

File: sqs/sqs.go (hunk 12 above)
Comment (andrewsykim) [reply sqs/sqs.go]: https://github.com/Wattpad/kube-sqs-autoscaler/pull/1/files#diff-53618cae32e23ae47e1c35f10da92654R143

File: main_test.go [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+func (m *MockSQS) SetQueueAttributes(input *sqs.SetQueueAttributesInput) (*sqs.SetQueueAttributesOutput, error) {
+	m.QueueAttributes = &sqs.GetQueueAttributesOutput{
+		Attributes: input.Attributes,
+	}
+	return &sqs.SetQueueAttributesOutput{}, nil
+}
+
+func NewMockSqsClient() *mainsqs.SqsClient {
+	Attributes := make(map[string]*string)
+	Attributes["ApproximateNumberOfMessages"] = aws.String("50")
```
Comment (jharlap) [inline main_test.go]: this seems like a good place to just use a map literal instead of `make` then assigning a single element.

File: main_test.go [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+
+type MockSQS struct {
+	QueueAttributes *sqs.GetQueueAttributesOutput
+}
+
+func (m *MockSQS) GetQueueAttributes(*sqs.GetQueueAttributesInput) (*sqs.GetQueueAttributesOutput, error) {
+	return m.QueueAttributes, nil
+}
+
+func (m *MockSQS) SetQueueAttributes(input *sqs.SetQueueAttributesInput) (*sqs.SetQueueAttributesOutput, error) {
```
Comment (jharlap) [inline main_test.go:222]: clearly you can never get an error from this call, so why return one? also, why not just accept an attributes map if that's the only part of the SetQueueAttributesInput object you're going to use? (remember that this method isn't here to satisfy an interface, it's here to let you control the mock during testing)

File: main_test.go [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+	assert.Equal(t, int32(*maxPods), deployment.Spec.Replicas)
+
+	Attributes = make(map[string]*string)
+	Attributes["ApproximateNumberOfMessages"] = aws.String("10")
+	input = &sqs.SetQueueAttributesInput{
+		Attributes: Attributes,
+	}
+	s.Client.SetQueueAttributes(input)
+
+	time.Sleep(10 * time.Second)
```
Comment (jharlap) [inline main_test.go:46]: this smells strongly. a) it means your test suite takes way longer to run than necessary, and b) it means that your tests are really sensitive to unexpected pauses/slowdowns on the test host so you're likely to have flakey tests.

File: main_test.go [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+	deployment, _ = p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(*minPods), deployment.Spec.Replicas)
+}
+
+func TestRunScaleUpCoolDown(t *testing.T) {
+	// override default vars for testing
```
Comment (jharlap) [inline main_test.go]: comments should tell us why the code does something, not what it does - we can read the code to know what it does. if the why is obvious (like this example), just omit the comment.

File: main_test.go [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+		Client: &MockSQS{
+			QueueAttributes: &sqs.GetQueueAttributesOutput{
+				Attributes: Attributes,
+			},
+		},
+		QueueUrl: "sqs.test.com",
+	}
+}
+
+func TestRunMaxMin(t *testing.T) {
```
Comment (jharlap) [inline main_test.go]: test func names are effectively documentation - when the test fails, if the name indicates what behaviour we're testing for then the test report output will be more useful. for example, this test name isn't conveying a lot of info, but i'd imagine it should verify that both the max and min replica counts are respected by the autoscaler...

File: main_test.go [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+	Attributes = make(map[string]*string)
+	Attributes["ApproximateNumberOfMessages"] = aws.String("10")
+	input = &sqs.SetQueueAttributesInput{
+		Attributes: Attributes,
+	}
+	s.Client.SetQueueAttributes(input)
+
+	time.Sleep(10 * time.Second)
+	deployment, _ = p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(*minPods), deployment.Spec.Replicas)
```
Comment (jharlap) [inline main_test.go]: when the test fails, if all you see is something like "TestRunMaxMin expected 1 got 2" (or whatever the assert library does to format its output) will that be informative?

File: main_test.go [hunk 20]
```diff
@@ -0,0 +1,224 @@
...
+	go Run(p, s)
+
+	Attributes := make(map[string]*string)
+	Attributes["ApproximateNumberOfMessages"] = aws.String("100")
+	input := &sqs.SetQueueAttributesInput{
+		Attributes: Attributes,
+	}
+	s.Client.SetQueueAttributes(input)
+
+	// with a 15 second wait and a 10 second cool down, we should expect only 1 pod to scale up
```
Comment (jharlap) [inline main_test.go]: is that true?
t=0 poll -> scale up
t=5 poll -> cool down noop
t=10.01 poll -> scale up (cool down has expired)

File: main_test.go [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+	time.Sleep(15 * time.Second)
+	deployment, _ := p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(4), deployment.Spec.Replicas)
```
Comment (jharlap) [inline main_test.go]: if we should only have had one pod scale up, how are we up to 4 replicas?

Comment (jharlap) [reply main_test.go]: oh, i see - we started at 3. that was very subtle 😕

File: scale/scale.go (hunk 2 above)
Comment (jharlap) [reply scale/scale.go]: if this TODO has been done, maybe remove the comment?

File: scale/scale.go [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+
+import (
+	"github.com/pkg/errors"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+// for testing purposes
```
Comment (jharlap) [inline scale/scale.go]: unnecessary comment. also, if you did want to have a comment on a type (which is always nice), it should be a godoc-appropriate comment.

File: scale/scale_test.go [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+		Max:        max,
+		Deployment: kubernetesDeploymentName,
+		Namespace:  kubernetesNamespace,
+	}
+}
+
+func TestScaleUp(t *testing.T) {
+	p := NewMockPodAutoScaler("test", "test", 5, 1)
+
+	_ = p.ScaleUp()
```
Comment (jharlap) [inline scale/scale_test.go]: when you explicitly discard the error like this, you really call attention to the fact that you should be checking the error ;P (btw, given there are two ways to address this, the one that checks the error is better than the one that hides the fact that you're discarding the error)

File: scale/scale_test.go [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+	deployment, _ := p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(4), deployment.Spec.Replicas)
+	_ = p.ScaleUp()
+	deployment, _ = p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(5), deployment.Spec.Replicas)
+
+	err := p.ScaleUp()
+	assert.NotNil(t, err)
+	deployment, _ = p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(5), deployment.Spec.Replicas)
```
Comment (jharlap) [inline scale/scale_test.go]: after a pause and contemplation, i can see why this is still 5 instead of 6, but here's where a "why" comment would make reading the code much faster. that comment should probably be a few lines up to explain this block as a whole

File: scale/scale.go [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+
+	currentReplicas := deployment.Spec.Replicas
+
+	if currentReplicas >= int32(p.Max) {
+		return errors.New("Max pods reached, not scaling up")
+	}
+
+	deployment.Spec.Replicas = currentReplicas + 1
+
+	// TODO: use same namespace that kube-sqs-autoscaler is using
```
Comment (jharlap) [inline scale/scale.go]: obsolete comment?

File: scale/scale.go [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+	_, err = p.Client.Deployments(p.Namespace).Update(deployment)
+	if err != nil {
+		return errors.Wrap(err, "Failed to scale up")
+	}
+
+	return nil
+}
+
+func (p *PodAutoScaler) ScaleDown() error {
+	// TODO: use the same namespace that kube-sqs-autoscaler is using
```
Comment (jharlap) [inline scale/scale.go]: looks like this can be made obsolete trivially now?

File: scale/scale_test.go [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+	p := NewMockPodAutoScaler("test", "test", 5, 1)
+
+	_ = p.ScaleDown()
+	deployment, _ := p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(2), deployment.Spec.Replicas)
+	_ = p.ScaleDown()
+	deployment, _ = p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(1), deployment.Spec.Replicas)
+
+	err := p.ScaleDown()
```
Comment (jharlap) [inline scale/scale_test.go]: again, why do we now care about the error?

File: scale/scale_test.go [hunk 28]
```diff
@@ -0,0 +1,113 @@
...
+	err := p.ScaleUp()
+	assert.NotNil(t, err)
+	deployment, _ = p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(5), deployment.Spec.Replicas)
+}
+
+func TestScaleDown(t *testing.T) {
+	p := NewMockPodAutoScaler("test", "test", 5, 1)
+
+	_ = p.ScaleDown()
```
Comment (jharlap) [inline scale/scale_test.go]: again, discarding error smell

File: sqs/sqs_test.go [hunk 29]
```diff
@@ -0,0 +1,47 @@
...
+
+type MockSQS struct {
+	QueueAttributes *sqs.GetQueueAttributesOutput
+}
+
+func (m *MockSQS) GetQueueAttributes(*sqs.GetQueueAttributesInput) (*sqs.GetQueueAttributesOutput, error) {
+	return m.QueueAttributes, nil
+}
+
+func (m *MockSQS) SetQueueAttributes(input *sqs.SetQueueAttributesInput) (*sqs.SetQueueAttributesOutput, error) {
```
Comment (jharlap) [inline sqs/sqs_test.go:27]: similar comments to those i made before - you can accept the attributes directly, and you don't need to return an error: this method only exists to make the mock controllable, not to satisfy an interface.

File: sqs/sqs_test.go [hunk 30]
```diff
@@ -0,0 +1,47 @@
...
+	Attributes := make(map[string]*string)
+	Attributes["ApproximateNumberOfMessages"] = aws.String("50")
+
+	return &SqsClient{
+		Client: &MockSQS{
+			QueueAttributes: &sqs.GetQueueAttributesOutput{
+				Attributes: Attributes,
+			},
+		},
+		QueueUrl: "sqs.test.com",
```
Comment (jharlap) [inline sqs/sqs_test.go]: when using example urls, you should really use example.com which is reserved by iana for exactly this purpose

File: sqs/sqs_test.go [hunk 31]
```diff
@@ -0,0 +1,47 @@
...
+func (m *MockSQS) SetQueueAttributes(input *sqs.SetQueueAttributesInput) (*sqs.SetQueueAttributesOutput, error) {
+	m.QueueAttributes = &sqs.GetQueueAttributesOutput{
+		Attributes: input.Attributes,
+	}
+
+	return &sqs.SetQueueAttributesOutput{}, nil
+}
+
+func NewMockSqsClient() *SqsClient {
+	Attributes := make(map[string]*string)
```
Comment (jharlap) [inline sqs/sqs_test.go:36]: one liner with a map literal would work nicely here

File: sqs/sqs_test.go [hunk 32]
```diff
@@ -0,0 +1,47 @@
...
+		Client: &MockSQS{
+			QueueAttributes: &sqs.GetQueueAttributesOutput{
+				Attributes: Attributes,
+			},
+		},
+		QueueUrl: "sqs.test.com",
+	}
+}
+
+func TestNumMessages(t *testing.T) {
```
Comment (jharlap) [inline sqs/sqs_test.go]: this feels a lot like you're only giving lip service to testing the method. are there really zero edge cases to test for?

File: scale/scale_test.go [hunk 33]
```diff
@@ -0,0 +1,113 @@
...
+	"github.com/stretchr/testify/assert"
+	"testing"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type MockDeployment struct {
```
Comment (jharlap) [inline scale/scale_test.go]: ok, now i see what you meant about the code duplication... this is maybe a bit more code than i would have duplicated, or at least you could put it in its own file so it doesn't clutter the real test file. if you really do want it in the same file, at least put it _after_ the tests so the most important code in the file is at the top.

File: main.go [hunk 34]
```diff
@@ -0,0 +1,82 @@
...
+			}
+		}
+	}
+
+}
+
+func main() {
+	flag.Parse()
+
+	p := scale.NewPodAutoScaler(*kubernetesDeploymentName, *kubernetesNamespace, *maxPods, *minPods)
```
Comment (jharlap) [inline main.go]: the reason i tend to avoid putting all the flag lines in the package-level var declaration block is because you're stuck with all these pointers instead of having nice clean values to use. you can really feel it in the test code where the number of asterisks is surely setting some kind of record ;P

File: sqs/sqs.go [hunk 35]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+type SqsClient struct {
```
Comment (jharlap) [inline sqs/sqs.go:19]: This name kinda stutters - from outside it's `sqs.SqsClient`. Also, even if `go vet` doesn't realize SQS is an acronym to complain about it, you should still make it all caps.

File: sqs/sqs.go [hunk 36]
```diff
@@ -0,0 +1,50 @@
...
+	Client   SQS
+	QueueUrl string
+}
+
+func NewSqsClient(queue string, region string) *SqsClient {
```
Comment (jharlap) [inline sqs/sqs.go:24]: more stutter - `sqs.NewSQSClient`? could just be `sqs.NewClient`, right?

**Review by jharlap — DISMISSED** [review]
- Have you tried running your tests with the race detector enabled?
- You haven't tested the cases where you start with more replicas than the max or fewer than the min... what should happen? what will happen?

**Comment by jharlap** [comment]
Also, seems you're missing a .travis.yml file?

File: main_test.go (hunk 16 above)
Comment (andrewsykim) [reply main_test.go:46]: Testing the Run function is more of an integration test. I agree that sleeping is a bit of a code smell. Any suggestions on alternatives?

File: main_test.go (hunk 19 above)
Comment (andrewsykim) [reply main_test.go]: Think its best to just separate the max and min case into two separate tests

File: main_test.go (hunk 20 above)
Comment (andrewsykim) [reply main_test.go]: the cool down is 10 seconds in this test case.

**Comment by andrewsykim** [comment]
@jharlap I addressed most of your comments, I don't want to keep this PR open for too long so I'll address the rest of your comments in another PR!
